x-pack/legacy/plugins/security/server/lib/authentication/providers/kerberos.ts

```diff
@@ -245,6 +215,54 @@ export class KerberosAuthenticationProvider extends BaseAuthenticationProvider {
     }
   }
 
+  /**
+   * This method is only called when authentication via access token stored in the state failed because of expired
+   * token. So we should use refresh token, that is also stored in the state, to extend expired access token and
+   * authenticate user with it.
+   * @param request Request instance.
+   * @param state State value previously stored by the provider.
+   */
+  private async authenticateViaRefreshToken(
+    request: RequestWithLoginAttempt,
+    state: ProviderState
+  ) {
+    this.debug('Trying to refresh access token.');
+
+    let refreshedTokenPair: TokenPair | null;
+    try {
+      refreshedTokenPair = await this.options.tokens.refresh(state.refreshToken);
+    } catch (err) {
+      return AuthenticationResult.failed(err);
+    }
+
+    // If refresh token is no longer valid, then we should clear session and redirect user to the
+    // login page to re-authenticate, or fail if redirect isn't possible.
+    if (refreshedTokenPair === null) {
+      this.debug('Both access and refresh tokens are expired. Re-initiating SPNEGO handshake.');
+      return this.authenticateViaSPNEGO(request, state);
+    }
+
+    try {
+      request.headers.authorization = `Bearer ${refreshedTokenPair.accessToken}`;
+
+      const user = await this.options.client.callWithRequest(request, 'shield.authenticate');
+
+      this.debug('Request has been authenticated via refreshed token.');
+      return AuthenticationResult.succeeded(user, refreshedTokenPair);
+    } catch (err) {
+      this.debug(`Failed to authenticate user using newly refreshed access token: ${err.message}`);
+
+      // Reset `Authorization` header we've just set. We know for sure that it hasn't been defined before,
+      // otherwise it would have been used or completely rejected by the `authenticateViaHeader`.
+      // We can't just set `authorization` to `undefined` or `null`, we should remove this property
+      // entirely, otherwise `authorization` header without value will cause `callWithRequest` to fail if
+      // it's called with this request once again down the line (e.g. in the next authentication provider).
+      delete request.headers.authorization;
```

Comment:
__note:__ I was about to delete all these `delete request.headers.authorization;` in `catch` and mutate request only if `_autenticate` succeeds using this new `headers` argument:
```ts
const user = await this.options.client.callWithRequest(request, 'shield.authenticate', {
  headers: { authorization }
});
```
But it doesn't allow to override header that is already in request, it's a bummer, we'll discuss this in the upcoming NP context-bound services RFC.